### ISSUE TITLE
Feature: Soft Delete

### ISSUE THREAD

**Issue by bartlomiej-korpus** [typeorm/typeorm#534]
I would like to know your opinion on implementing Soft Delete in TypeORM.
I would love to help doing this.

**Comment by pleerock** [comment]
need design proposal:

* we need define a column which will store state of deletion. How user should specify this column?
* when we call `repository#remove(entity)` we mark entity as removed instead of removing it from db. Or we shall create a separate method for that?
* when we call finds and use query builder we automatically add `where ... and entity.isDeleted = false`. what about joins subselects and other operations? This seems to be complex need to think this question more.
* how do we forcely delete object from a db?

**Comment by bartlomiej-korpus** [comment]
I will do some research on how this is implemented in other ORMs and come back with a proposal.

**Comment by NoNameProvided** [comment]
I would love to see this feature in, currently we filter for this in our services which works well but not so elegant.

**Comment by pleerock** [comment]
@bartlomiej-korpus  any news on a proposal?

**Comment by rajshrimohanks** [comment]
I would totally love this feature as well. I used Sequelize before TypeORM and their implementation is such that they have a `deletedAt` column along with `createdAt` and `updatedAt`. If the `deletedAt` column is populated, the record is considered as soft delete. If not (empty), then the record is considered as not deleted. Their `find()` method doesn't return soft-deleted records by default, but can be forced with a flag.

Currently in TypeORM, I'm filtering manually, but would love to see a built in way to achieve this.

Thanks again for an awesome library! :+1:

**Comment by pleerock** [comment]
Yeah I think we can go same way

**Comment by pleerock** [comment]
BTW this is not a hard to implement feature and opened for community for those who want to contribute to typeorm.

**Comment by NoNameProvided** [comment]
Shouldn't it be better to add it as an option to `find`, etc? Having a different call for all makes the API more verbose than it should be.

**Comment by rajshrimohanks** [comment]
I would like it to be a an option for `find` as well. Instead of adding three more methods, it'll simply be an option. Would help in building cleaner code too.

**Comment by mrkmg** [comment]
I was thinking about that. Should I just added another overloaded method? 

Would it be best to just to add a parameter to the FindManyOptions and the FineOneOptions? Something like: withDeleted: boolean?

Thoughts?

**Comment by NoNameProvided** [comment]
We already have `find` and `findMany` options objects, those should be extended.

**Comment by mrkmg** [comment]
Yeah, those options are the `FindOneOptions` and `FineManyOptions` objects. I will add to those.

Does the key `withDeleted: boolean` sound good?

**Comment by rajshrimohanks** [comment]
Sounds good for me. :)

**Comment by codeandcats** [comment]
Hey guys, what's the progress on this?

I'd love to be using this myself.

**Comment by codeandcats** [comment]
Also, if we've decided that soft delete belongs in the ORM and not in a service layer, then I think it would be worth considering adding options for having `createDate` and `updateDate` fields as well. Though this can be a separate issue and PR, just thought I'd mention it here since it's related.

**Comment by NoNameProvided** [comment]
There is already `CreateDateColumn` and `UpdateDateColumn` decorator for this. You can use it like: 

```ts
  @CreateDateColumn({ type: 'timestamp' })
  public createdAt: Date;

  @UpdateDateColumn({ type: 'timestamp' })
  public updatedAt: Date;
```

**Comment by codeandcats** [comment]
Excellent! 👍

My apologies, I did look through the doco but somehow missed it.

**Comment by kikar** [comment]
This is a very useful feature we use a lot! Thank you for implementing.
As a suggestion, we need sometimes the default value of the column deletedAt as 0 instead of NULL, because MYSQL does not work well will UNIQUE indexes that include nullable columns.
So the `@SoftDeleteDateColumn()` should accept a type and default value, and the find methods should check for this default to be excluded.

**Comment by knnth** [comment]
I hope this feature can be possible in the next releases! It's useful for many projects..

**Comment by codeandcats** [comment]
Just checking if any progress has been made on this?

I’m currently implementing `deletedAt` manually.

**Comment by tranghaviet** [comment]
It's great if TypeORM have soft delete feature. Laravel Eloquent do it very well.

**Comment by oeddyo** [comment]
would love to see this soon too!

**Comment by gregwym** [comment]
We are implementing this ourselves by extending our own Custom repository for everything. However, it cannot handle soft deleted model for relations. Would love to see progress on this and how other people are doing it.

**Comment by carlosbensant** [comment]
@gregwym 

I thought it should be possible for users of esoteric platforms to override the default implementation with their own custom implementations as well, whether in the service layer or like you in a custom repository but I hadn't thought that part through "how it'd handle soft-deleted relation entities". I'd like to see how did you end up setting it up after all.

**Comment by mazyvan** [comment]
What about retrieving data from a view table that only knows about non deleted values. And make every selection, including joins should use that view

**Comment by MouaadAitSaid** [comment]
Guys any progress has been made on this?
I’m actually implementing deletedAt manually in my code, but it's not smooth and elegant..

**Comment by robertmain** [comment]
Just adding my two cents to this. This is one of my service methods right now:


```typescript
  public async findAll(includeDeleted: boolean = false): Promise<Entity[]> {
    const where = { deletedAt: null };
    if (includeDeleted) {
      delete where.deletedAt;
    }
    const ingredients = await this.repository.find({
      where,
    });
    return ingredients;
  }
```

**Comment by haschu** [comment]
I would love to see this soon as well and I'm going to ship a PR in about 2-3 weeks (or maybe sooner if I find some time).

Please let me know if anybody is actively working on this feature, so I wouldn't waste my time on this :)

**Comment by jonathanrdelgado** [comment]
@haschu https://github.com/typeorm/typeorm/pull/1192

**Comment by haschu** [comment]
@jonathanrdelgado Wow, thanks! No idea why I did not find this myself...

**Comment by bellstrand** [comment]
@haschu You should check what needs to be done before such a PR could be merged.
If I recall the conversation, @pleerock was planing on refactoring the QueryBuilder, and this functionality was put on hold.
My teams are also planning on forking TypeORM and implementing a SoftDelete. It would be great to know what we need to complete before a SoftDelete PR would be considered.

**Comment by robertmain** [comment]
@bellstrand I mean that PR Is also full of merge conflicts too.....which will obviously need to be addressed before merge....last action on that was 22nd Jan so I don't know what the timeline for it is...

**Comment by bellstrand** [comment]
@robertmain I'm well aware of that. My thought was that we should check what needs to be done before fixing all those issues so that it can be merged in this time.

**Comment by robertmain** [comment]
Yep. And *I'm* just saying that it might be a while before it gets merged :)

**Comment by haschu** [comment]
@bellstrand I'm aware of this, but thanks for pointing out. I just didn't find the PR, so I wasn't even aware that someone actually started working on it :P 

We have to wait until @pleerock got some more time to refactor the `QueryBuilder`.  Until then, soft-delete can be implemented via a service wrapper (that's what we do atm).

**Comment by PachowStudios** [comment]
@haschu Would you be willing to share your wrapper for handling soft deletes? Currently, we're using an extension of BaseEntity I wrote.

Here's what I'm doing:
```typescript
import { Field } from 'type-graphql';
import { BaseEntity, BeforeRemove, Column, EntityManager, FindManyOptions, FindOneOptions } from 'typeorm';

export abstract class SoftDeletableEntity extends BaseEntity {
  @Field()
  @Column({ default: true })
  isActive!: boolean;

  @BeforeRemove()
  beforeRemove() {
    throw new Error(`Attempted to hard-delete a soft-deletable ${this.constructor.name}`);
  }

  async softRemove(): Promise<void> {
    await this.softRemoveInternal();
  }

  async softRemoveUsing(manager: EntityManager): Promise<void> {
    await this.softRemoveInternal(manager);
  }

  private async softRemoveInternal(manager?: EntityManager): Promise<void> {
    const thisClass = this.constructor as typeof SoftDeletableEntity;

    if (!this.isActive) {
      throw new Error(`Attempted to soft-delete a(n) ${thisClass.name} that has already been soft-deleted`);
    }

    manager = manager || thisClass.getRepository().manager;
    await manager.update(thisClass, thisClass.getId(this), { isActive: false });
  }

  static async count<TStatic extends typeof BaseEntity, T extends InstanceType<TStatic>>(
    this: TStatic,
    options?: FindManyOptions<T>,
  ): Promise<number> {
    return super.count<T>({
      where: { isActive: true },
      ...options,
    });
  }

  static async find<TStatic extends typeof BaseEntity, T extends InstanceType<TStatic>>(
    this: TStatic,
    options?: FindManyOptions<T>,
  ): Promise<T[]> {
    return super.find<T>({
      where: { isActive: true },
      ...options,
    });
  }

  static async findAndCount<TStatic extends typeof BaseEntity, T extends InstanceType<TStatic>>(
    this: TStatic,
    options?: FindManyOptions<T>,
  ): Promise<[T[], number]> {
    return super.findAndCount<T>({
      where: { isActive: true },
      ...options,
    });
  }

  static async findByIds<TStatic extends typeof BaseEntity, T extends InstanceType<TStatic>>(
    this: TStatic,
    ids: string[],
    options?: FindManyOptions<T>,
  ): Promise<T[]> {
    return super.findByIds<T>(ids, {
      where: { isActive: true },
      ...options,
    });
  }

  static async findOne<TStatic extends typeof BaseEntity, T extends InstanceType<TStatic>>(
    this: TStatic,
    id?: string,
    options?: FindOneOptions<T>,
  ): Promise<T | undefined>;
  static async findOne<TStatic extends typeof BaseEntity, T extends InstanceType<TStatic>>(
    this: TStatic,
    options?: FindOneOptions<T>,
  ): Promise<T | undefined>;
  static async findOne<TStatic extends typeof BaseEntity, T extends InstanceType<TStatic>>(
    this: TStatic,
    idOrOptions?: string | FindOneOptions<T>,
    maybeOptions?: FindOneOptions<T>,
  ): Promise<T | undefined> {
    const id = idOrOptions instanceof Object ? undefined : idOrOptions;
    const options: FindOneOptions<T> = {
      where: { isActive: true },
      ...idOrOptions instanceof Object ? idOrOptions : maybeOptions,
    };

    return super.findOne<T>(id, options);
  }

  static async findOneOrFail<TStatic extends typeof BaseEntity, T extends InstanceType<TStatic>>(
    this: TStatic,
    id?: string,
    options?: FindOneOptions<T>,
  ): Promise<T | undefined>;
  static async findOneOrFail<TStatic extends typeof BaseEntity, T extends InstanceType<TStatic>>(
    this: TStatic,
    options?: FindOneOptions<T>,
  ): Promise<T | undefined>;
  static async findOneOrFail<TStatic extends typeof BaseEntity, T extends InstanceType<TStatic>>(
    this: TStatic,
    idOrOptions?: string | FindOneOptions<T>,
    maybeOptions?: FindOneOptions<T>,
  ): Promise<T | undefined> {
    const id = idOrOptions instanceof Object ? undefined : idOrOptions;
    const options: FindOneOptions<T> = {
      where: { isActive: true },
      ...idOrOptions instanceof Object ? idOrOptions : maybeOptions,
    };

    return super.findOneOrFail<T>(id, options);
  }
}
```

**Comment by haschu** [comment]
@PachowStudios Our service wrapper isn't finished yet. I'm going to share it, when it's ready :)

But basically it's pretty similar to your example, despite we're using repositories instead of active directory and instead of your `isActive` boolean we're using a `deletedAt` `Date` column.
So we're able to do something like "hard delete all soft deleted records that have been deleted for more than 30 days".

Also we introduced a new directive `@DeletedAt` directive, to marks the property/column that stores the `deletedAt` `Date`.

Hope that gives you some ideas, but your solution is also pretty neat :)

**Comment by aaldrin23** [comment]
Any update on this?

**Comment by mswezey23** [comment]
Any ETA?

**Comment by mrkmg** [comment]
AFAIK, this feature is on hold until a re-write of some of the internals of TypeORM are complete.

**Comment by johinsDev** [comment]
@haschu Can you share a code for create the decorator @DeletedAt?

**Comment by haschu** [comment]
@johinsDev I'm going to share it, when it's ready :)

Our team is currently implementing a TypeORM extension wrapper including functionality described in #4280 (soft deletion is included as well). We're willing to publish it here on GitHub or as a npm-package if people are interested.
I cannot give a specific time though, as this project currently has a minor priority for us (I'd guess 2-3 weeks, maybe earlier).

**Comment by nolazybits** [comment]
@johinsDev I'll be keen thanks 👍

**Comment by PachowStudios** [comment]
@haschu I would be really interested to see what ya'll have so far! If you'd be willing to share your WIP maybe we can bounce some ideas off each other.

**Comment by jpollard-cs** [comment]
@haschu would be interested to see that - also curious if it's compatible with cascades?

**Comment by haschu** [comment]
Hey, sorry for the late response - I'm extremely busy atm.

@jpollard-cs Sorry, it won't be compatible with cascades.

@PachowStudios Our current WIP is tied to in-house framework code that I cannot share. But as it seems that some people are interested, I'm going to outsource it into a GitHub repository soon. I'll keep you up to date!

**Comment by jpollard-cs** [comment]
@pleerock you indicated there were plans to rewrite `QueryBuilder` before support for soft deletes should be added (https://github.com/typeorm/typeorm/pull/1192#issuecomment-456294376) has this been done?

**Comment by ksharonov** [comment]
any news?

**Comment by haschu** [comment]
@ksharonov Working on it. Have to finish a project first, which takes a bit longer than excepted.

**Comment by vlaktion** [comment]
Yes it's harder than it sounds

**Comment by jpollard-cs** [comment]
from @pleerock via TypeORM slack:
> this feature doesn’t fit well with exist functionality we have. consider your own implementation on domain-level

**Comment by dantman** [comment]
I did a comparison of ORMs across various languages to get an idea of how standard support for some form of soft deletion (also sometimes called paranoia or safe delete).

Some of them bake it into the library. Some of them have a strong plugin system and soft delete is implemented in an official plugin, a widely known plugin, and sometimes just a plugin a 3rd party has made (IIRC the plugin is (semi-)official in most of the ORMs that would be considered mature). One doesn't implement soft delete per-se, but implements model annotations that override the behaviour of a delete operation and default queries so it's easy to apply soft delete behaviours with official annotations. Some don't support soft delete but have hooks for all the built in queries/etc so it's easy to write your own soft delete without wrapping the whole API. And one also officially implements an "archive" feature where instead of setting a `deleted_at`/`is_deleted` column it deletes the row and inserts a copy into an archive table.

TypeORM does appear to be the outlier, being pretty much the only "mature" ORM that does not have soft delete built in and does not have a plugin or hook system that allows soft delete to be implemented outside of the ORM without wrapping the API.

Language | ORM              | Supports soft delete?
-------- | ---------------- | ---------------------
Ruby     | ActiveRecord     | ✅yes (plugin)
Ruby     | DataMapper       | ✅yes
Ruby     | Sequel           | ✅yes (plugins)
PHP      | Doctrine ORM     | ✅yes (official plugin)
PHP      | Propel           | ✅yes
PHP      | Eloquent/Laravel | ✅yes
PHP      | RedBean          | ❌no (✔️plugins)
Java     | Hibernate        | ✅yes/sortof (annotations)
Java     | Ebean ORM        | ✅yes
Python   | Django           | ✅yes (plugins)
Python   | SQLAlchemy       | ✅yes (plugins)
ES       | Sequelize        | ✅yes
ES       | Bookshelf        | ✅yes (plugin)
ES       | Waterline        | ⚠️no (✅archive, ✅bespoke soft delete)
ES       | Objection        | ✅yes (plugin)
ES       | TypeORM          | ❌no (❌bespoke)

**Comment by jpollard-cs** [comment]
> from @pleerock via TypeORM slack:
> 
> > this feature doesn’t fit well with exist functionality we have. consider your own implementation on domain-level
> 

@pleerock is it possible to get an explanation why this functionality does not fit well with the existing functionality? It would be beneficial to the community to understand the theory here. We have a number of folks who are interested in such functionality and willing to contribute (including myself). I'm sure we are all extremely grateful for the hard work you have put into this project (I know I am) and it's great to see a TypeScript-first ORM library that supports the repository pattern and not just the active record pattern. If there's not a path to natively supporting this is there at least a path to making TypeORM more modular and extensible so features like this can more easily be supported by the community?

**Comment by iWinston** [comment]
Hello, I have done some work to support soft-delete and query scopes on a fork of TypeORM. I name it TypeORM+(https://github.com/iWinston/typeorm-plus), and it works fine in our project. 
There are some features ：
1. @DeleteDateColumn() @DeleteDateColumn is a special column that is automatically set to the entity's delete time each time you call delete of entity manager or repository softly. The default value of it is null.
2. Soft Delete API for Repository:
-  `softDelete` - Deletes entities by entity id, ids or given conditions softly.
- `restore ` - Restore entities by entity id, ids or given conditions.
- `softRemove` - Removes a given entity or array of entities softly.
- `recover` - Recover a given entity or array of entities.
3. Soft Delete API for QueryBuilder:
- `softDelete ` - Create soft DELETE queries.
- `restore` - Create queries to restore objects from a db.
4. Query scopes allow you to add constraints to all queries for a given entity. If the @DeleteDateColumn is set, the default scope will be "non-deleted". So it will only pull "non-deleted" entities from the database.
- You can apply scopes to Repository using the option of scope
- You can apply scopes to QueryBuilder using the method `setScope`
5. The soft delete is an update actually, so it will broadcast an update event. The soft delete hasn't supported subselects yet.

I hope I can receive some advice from you. And I would like to make a PR, but I am not sure if someone was doing the same thing? And should I wait until @pleerock refactored the QueryBuilder?

**Comment by PachowStudios** [comment]
@iWinston Wow this is really awesome! Hopefully these improvements can be merged in the near future.

I really like the concept of scopes, it's a smart way to solve the problem of having default sets of conditions. Is that your own idea, or did you pull that from another ORM?

**Comment by GarryOne** [comment]
@PachowStudios the same concept of Query Scopes exists in Laravel's ORM: Eloquent.
https://laravel.com/docs/5.8/eloquent#query-scopes

**Comment by JClackett** [comment]
@PachowStudios and in Rails' ActiveRecord

**Comment by iWinston** [comment]
Thank you for your love of TypeORM+ @PachowStudios.

Yes, I learned something from the Eloquent/Laravel at the beginning of building TypeORM+. And I tried to use the scopes in  a more `TypeORM` style with the `FindConditions`.

**Comment by apollonin** [comment]
I see that soft deletion will be available in 0.3.0
https://typeorm.io/#/roadmap/030

Has someone tried it already?

**Comment by iWinston** [comment]
@apollonin I have made a PR: #5034.

**Comment by jpollard-cs** [comment]
@pleerock is this true about 0.3? Will this use the work @iWinston’s been doing?

**Comment by rafaelfesi** [comment]
According to the [roadmap](https://typeorm.io/#/roadmap/030), the feature of soft deletion is planned for version 0.3.0.

**Comment by jpollard-cs** [comment]
Yeah @rafaelfesi just surprised as we were previously informed that this would not be supported

**Comment by Destreyf** [comment]
I just want to comment here @rafaelfesi, @jpollard-cs, and @apollonin.

If you look at the history for the road map, (based on what is available in GitHub) it hasn't been updated since Apr 16, 2018, there was a point where it looks like the contents were copied out then back in, but as far as the actual commits/changes it hasn't been updated in over a year.

You can see the commits here:
https://github.com/typeorm/typeorm/commits/master/docs/roadmap.md

At this point, I highly doubt soft deletion is intended to be in the 0.3.0 release, and based on prior comments like this one: https://github.com/typeorm/typeorm/issues/534#issuecomment-523439829

There are also two separate pull requests open currently, with @iWinston's being the newer one, but there doesn't seem to be any indication (yet) of using it.

This is a feature that i am sorely in need of, however based on the current outlook, I have resorted to building my own base repository and replacing/extending the functionality that was provided by TypeORM, this allows me to leverage TypeORM directly, without needing to use forks of the project, but it's not an elegant solution and has resulted in quite a bit of duplicate code.

**Comment by hselaaron** [comment]
Any updates here? Thanks 😊

**Comment by pleerock** [comment]
Let's bring all your thoughts in #5034, review and try to merge that PR. Thanks for patience.

**Comment by fwpushan** [comment]
Any suggestion/planning/design, how to handle the relationships with soft delete?

**Comment by lgarczyn** [comment]
Currently `insert` does not work over a soft-deleted entity, which makes some sense.

`save` does work, but there's still no options in save to set this behavior, I think such options should look like:

// If entity does not exist
* onInsert: 'continue' | 'fail'
// If entity exists
* onUpdate: boolean | 'continue' | 'erasePrevious' | 'updatePrevious' | 'fail'
// If entity is deleted
* onDeleted: boolean | 'continue' | 'erasePrevious' | 'updatePrevious' | 'fail'

**Comment by IonelLupu** [comment]
The BaseEntity still needs those `soft*` methods

**Comment by ueharajohji** [comment]
how to get data who got soft removed?

`deleteAt is not null`

**Comment by lgarczyn** [comment]
@ueharajohji absolutely, assuming you don't intend specific behavior from future deleteAt values

**Comment by ueharajohji** [comment]
@lgarczyn 

Assuming typeorm is by default getting data with this condition

`deleteAt is null`

adding `deleteAt is not null` is a little bit off to me. are there any better ways to do 
what I am trying to achieve?

**Comment by imnotjames** [comment]
Soft delete support was merged - given that, I'm going to close this issue.

> The BaseEntity still needs those soft* methods as the Repository

Please see and support https://github.com/typeorm/typeorm/issues/6720

If you have other issues please open them separately as there's a lot of context here that's no longer relevant.

For questions, please check out the [community slack](https://join.slack.com/t/typeorm/shared_invite/zt-gej3gc00-hR~L~DqGUJ7qOpGy4SSq3g) or check [TypeORM's documentation page on other support avenues](https://typeorm.io/#/support) - cheers!

**Comment by biels** [comment]
Is this feature documented somewhere?

**Comment by ford-sennalabs** [comment]
@biels 
I didn't find in documented explicitly 

but @iWinston's comments is very useful and detailed.

**Comment by bellstrand** [comment]
@biels It's spread out. This commit contains most of it.
https://github.com/typeorm/typeorm/commit/5b50f474034a570d3e209e15a6dc8f25fab73739

**Comment by trungtran1** [comment]
@bartlomiej-korpus https://github.com/typeorm/typeorm/issues/534#issue-235066314
But now I using the find() method, it still returns entities with deletedAt field not null.
I really only want to return entities that have not been deleted (ie deletedAt is null)

**Comment by dungpv10** [comment]
Have any changed ? Have any like : deletedAt column, forceDelete, findWithTrash ?
